lib/awful/widget/watch.lua

```diff
@@ -60,22 +62,48 @@ local watch = { mt = {} }
 -- For "exit" reason it's the exit code.
 -- For "signal" reason — the signal causing process termination.
 --
--- @param[opt=wibox.widget.textbox()] base_widget Base widget.
+-- @tparam[opt=wibox.widget.textbox()] widget base_widget Base widget.
 --
 -- @return The widget used by this watch
-function watch.new(command, timeout, callback, base_widget)
-    timeout = timeout or 5
+function watch.new(command, interval, callbacks, base_widget)
+    interval = interval or 5
     base_widget = base_widget or textbox()
-    callback = callback or function(widget, stdout, stderr, exitreason, exitcode) -- luacheck: no unused args
-        widget:set_text(stdout)
+
+    if type(callbacks) == 'function' then
+        callbacks = {success=callbacks}
+    elseif not callbacks then
+        callbacks = {
+            success = function(widget, stdout, _)
+                widget:set_text(stdout)
+            end
+        }
+    end
+
+    local notification_id
+    callbacks.error = callbacks.error or function(stdout, stderr, exitcode, type)
+        -- widget:set_text(stderr)
+        local output = util.concat_table({stdout=stdout, stderr=stderr}, "\n") or "No output"
+        local title = (type == 'error'
+            and "An error occurred with a watch widget. Stopping it."
+            or  "A signal was received for a watch widget. Stopping it.")
+        local text = string.format(
+            "%s\n<b>%s:</b> %s",
+            output, (type == 'error' and "Exit code" or "Signal"), exitcode)
+        notification_id = naughty.notify({preset = naughty.config.presets.critical,
+                        title = title,
+                        text = text,
+                        replaces_id = notification_id,
+                    }).id
+        base_widget:set_text(stdout)
     end
-    local t = timer { timeout = timeout }
+    callbacks.signal = callbacks.signal or callbacks.error
```

Comment:
> So, if the caller provides an error callback, it will also be used for signals? That seems surprising.

Let's use it then only when no explicit error handler was used?!